### ISSUE TITLE
CORS doesn't work with `DependsOnMethod`

### ISSUE THREAD

**Issue by leosuncin** [RobinTail/express-zod-api#705]
I was trying `express-zod-api` and found errors with CORS.

![imagen](https://user-images.githubusercontent.com/4307697/203210766-0f69be62-acd1-4740-b2b9-31e25c5d68f0.png)

For request with methods different of GET and POST, it looks like the CORS middleware doesn't exits, similar to #514

Reproduction repository https://github.com/leosuncin/express-zod-api-cors-error

**Comment by RobinTail** [comment]
Hello @leosuncin ,

I'm sorry, but your repo is relatively big... so I'm not sure I figured out what's happening there.
But, do I understand correctly, that on your screenshot you're calling a path that is not present in the [routing](https://github.com/leosuncin/express-zod-api-cors-error/blob/master/apps/api/src/routing.ts)?

I see you have `/health` and `/todo` specified, but you're calling `localhost:3000/...` just with some hash or ID. Right?

So, could you please clarify how this request you expect to be handled? Since it appears to me as you're requesting the path that does not listed in the routing, so therefore it does not have any supported methods at all.

Looking forward to your reply.

P.S. It would be great if you could provide me with exact `cURL` command to reproduce the issue. So I could clone your repo, start it and check it out.

**Comment by leosuncin** [comment]
Sorry for the mess, I simplified and reduced the code a lot, please check it again.

A `cURL` command won't help to debug it since the CORS issue only happens when the browser does a Cross-Origin Request, that's the reason why the reproduction code is composed by a server and a front-end app running on different ports.

**Comment by leosuncin** [comment]
Furthermore, I encountered other errors: the `Authorization` header is not accepted because the default configuration only allow `Content-Type`

https://github.com/RobinTail/express-zod-api/blob/2385f16040116f40c4972c080f9f75577dc0a25c/src/endpoint.ts#L201

This also cause problems with cookies because the header `Set-Cookie` is not available in the client side.

I'd like to recommend to use [cors](https://www.npmjs.com/package/cors) instead of defining the headers

https://github.com/RobinTail/express-zod-api/blob/45f34af4649ca0814bba14b430601afd765766ec/src/endpoint.ts#L193-L203

**Comment by RobinTail** [comment]
@leosuncin 

> Furthermore, I encountered other errors: the Authorization header is not accepted because the default configuration only allow Content-Type
> This also cause problems with cookies because the header Set-Cookie is not available in the client side.

It's already possible.
For that you need to change the config setting from `cors: true` to a function that returns headers.
Please check out the following article of the documentation:

https://github.com/RobinTail/express-zod-api#cross-origin-resource-sharing

> I'd like to recommend to use [cors](https://www.npmjs.com/package/cors) instead of defining the headers

Check out the following example on how to connect the `cors` express middleware:
https://github.com/RobinTail/express-zod-api#using-native-express-middlewares

**Comment by RobinTail** [comment]
@leosuncin 

> A cURL command won't help to debug it since the CORS issue only happens when the browser does a Cross-Origin Request, that's the reason why the reproduction code is composed by a server and a front-end app running on different ports.

Any request is possible to reproduce.
In particular I'm asking you to copy `OPTIONS` request as `cURL` from your browser or a tool you're using.
This option is available in a context menu.

**Comment by leosuncin** [comment]
The `OPTIONS` request

```sh
curl 'http://localhost:3010/hello' -X OPTIONS -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:107.0) Gecko/20100101 Firefox/107.0' -H 'Accept: */*' -H 'Accept-Language: es-MX,es;q=0.8,en-US;q=0.5,en;q=0.3' -H 'Accept-Encoding: gzip, deflate, br' -H 'Access-Control-Request-Method: PUT' -H 'Referer: http://localhost:3020/' -H 'Origin: http://localhost:3020' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-site' -H 'Pragma: no-cache' -H 'Cache-Control: no-cache'
```

The `PUT` request
```sh
curl 'http://localhost:3010/hello' -X PUT -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:107.0) Gecko/20100101 Firefox/107.0' -H 'Accept: */*' -H 'Accept-Language: es-MX,es;q=0.8,en-US;q=0.5,en;q=0.3' -H 'Accept-Encoding: gzip, deflate, br' -H 'Referer: http://localhost:3020/' -H 'Origin: http://localhost:3020' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-site'
```

Both works in the terminal, but not in the browser

Download [CORS_error.zip](https://github.com/RobinTail/express-zod-api/files/10096427/CORS_error.zip), unzip it and upload the HAR file to https://toolbox.googleapps.com/apps/har_analyzer/

**Comment by RobinTail** [comment]
Thank you @leosuncin .

For some reason your setup of the `/hello` endpoint returns the following header in the response to `OPTIONS` request:

```
Access-Control-Allow-Methods: GET, OPTIONS
```

Investigation continues.

**Comment by RobinTail** [comment]
Yeah, this is a very nice bug, @leosuncin .
Thank you very for finding this!

The reason is the way of processing the `DependsOnMethod` instance. In your case each method technically has a separate `Endpoint` instance. Each of them is for one method only. The current implementation creates several `OPTIONS` handlers for the same path, but only the first one being called (GET). So it responds with the wrong list of methods actually supported by the path.

This is bug. I'm going to find the way to fix it.

**Comment by RobinTail** [comment]
The exact location of the issue:
https://github.com/RobinTail/express-zod-api/blob/45f34af4649ca0814bba14b430601afd765766ec/src/routing.ts#L66

**Comment by RobinTail** [comment]
The issue with allowed methods is fixed in recently released v8.4.1 🚀 
Thank you again for finding this, @leosuncin 
I'm also going to fix it in v7.

It should work in your repo now, but without authorization.
In order to allow `Authorization` header, consider changing `cors` config option into a function. See the example above.
